### PR TITLE
docs(tools-dgeni, daffio): generate links through dgeni

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
- Dgeni now generates anchor tags from tags of the format "{@link product/DaffProductFacade#product$ the product$ field}".
- Adds a fragment link to every property of every component/class/etc so that they can be navigated to via the url.
- Stops angular from sanitizing the generated docs, so that the fragment links are not removed. I think this should be fine, since this is just documentation.
- Checks the url in daffio for a fragment. If the fragment exists, it will scroll down to that fragment link. I tried to do this in a less "native" way, but was unable to get it to work any other way. It works well as is anyway. I have an example I added in the product package in this PR.

### CODE REVIEW

File: apps/daffio/src/app/docs/shared/services/docs.service.spec.ts [hunk 1]
```diff
@@ -40,7 +40,19 @@ describe('DaffioDocService', () => {
     expect(req.request.method).toEqual('GET');
 
     req.flush(doc);
-  });
+	});
+
+  it('should be able to retrieve a doc when there is a fragment in the url', () => {
+    service.get('my/path#fragment').subscribe((apiDoc) => {
+      expect(apiDoc).toEqual(doc);
+    });
```
Comment (griest024) [inline apps/daffio/src/app/docs/shared/services/docs.service.spec.ts:48]: can you not use `jasmine-marbles` for this expectation?

File: tools/dgeni/src/processors/createAnchorTags.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+function buildAnchorTag(linkString): string {
+	let splitValues = linkString.split(' ');
+	//the title is always all of the values beyond the second part of the split values.
+	let title = splitValues.slice(2).join(' ');
+	//remove the } from the end of the title string.
+	title = title.substring(0, title.length - 1);
+	//build and return the anchor tag.
+	return '<a href=\'' + API_DOCS_PATH + splitValues[1] + '\'>' + title + '</a>';
```
Comment (griest024) [inline tools/dgeni/src/processors/createAnchorTags.ts:36]: I think you should use template strings here.

File: tools/dgeni/src/processors/createAnchorTags.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		})
+
+		return docs;
+	}
+}
+
+function buildAnchorTag(linkString): string {
+	let splitValues = linkString.split(' ');
+	//the title is always all of the values beyond the second part of the split values.
+	let title = splitValues.slice(2).join(' ');
```
Comment (griest024) [inline tools/dgeni/src/processors/createAnchorTags.ts:32]: I think you could probably use regex capture groups to get these values a little more cleanly.

**Review by griest024 — CHANGES_REQUESTED** [review]
Do you know why descriptions aren't showing up?

**Comment by griest024** [comment]
> Are they not? I see them

They aren't for me. Let me rebuild everything and try again

**Comment by griest024** [comment]
@lderrickable a full rebuild fixed the issue.

File: apps/daffio/src/app/docs/shared/components/doc-viewer/doc-viewer.component.ts [hunk 4]
```diff
@@ -1,16 +1,42 @@
...
+	}
+
+	ngOnInit() {
+		this.route.fragment.subscribe(fragment => { this.fragment = fragment; });
+	}
 
+  ngAfterViewInit() {
+		//scroll to the fragment in the routing url if it exists.
+		const fragmentElement = document.querySelector('#' + this.fragment);
+		if(fragmentElement) window.scrollTo({top: fragmentElement.getBoundingClientRect().top-FRAGMENT_SCROLL_OFFSET, behavior: 'smooth'});
```
Comment (lderrickable) [inline apps/daffio/src/app/docs/shared/components/doc-viewer/doc-viewer.component.ts:40]: I'm going to redo how this part works after talking to Damien about some things

File: tools/dgeni/src/processors/createAnchorTags.ts (hunk 3 above)
Comment (lderrickable) [reply tools/dgeni/src/processors/createAnchorTags.ts]: changed

File: tools/dgeni/src/processors/createAnchorTags.ts (hunk 2 above)
Comment (lderrickable) [reply tools/dgeni/src/processors/createAnchorTags.ts]: changed

File: apps/daffio/src/app/docs/shared/services/docs.service.spec.ts (hunk 1 above)
Comment (lderrickable) [reply apps/daffio/src/app/docs/shared/services/docs.service.spec.ts]: changed

File: apps/daffio/src/app/docs/shared/components/doc-viewer/doc-viewer.component.ts (hunk 4 above)
Comment (griest024) [reply apps/daffio/src/app/docs/shared/components/doc-viewer/doc-viewer.component.ts:40]: What was the change here? Shouldn't browsers natively support scrolling to the element with that ID?

File: tools/dgeni/src/processors/createAnchorTags.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+		})
+
+		return docs;
+	}
+}
+
+function buildAnchorTag(linkString: string): string {
+	const path = linkString.match(INLINE_LINK_PATH_REGEX)[0];
+	const title = linkString.match(INLINE_LINK_TITLE_REGEX)[0];
+	return `<a href='` + API_DOCS_PATH + path + `'>` + title + `</a>`;
```
Comment (griest024) [inline tools/dgeni/src/processors/createAnchorTags.ts:34]: With template strings you can interpolate values directly into it. It makes it easier to see exactly how the final string will look. So:
```ts
`<a href='${API_DOCS_PATH}${path}'>${title}</a>`
```

It looks kinda similar to inserting values in Angular templates. I think its far easier to read.

**Review by griest024 — CHANGES_REQUESTED** [review]

File: tools/dgeni/src/processors/createAnchorTags.ts (hunk 5 above)
Comment (lderrickable) [reply tools/dgeni/src/processors/createAnchorTags.ts]: oh ok. changed

**Review by griest024 — DISMISSED** [review]

**Pull request closed without merging**